File: composeApp/src/androidMain/kotlin/com/alextos/thousand/data/repository/mappers/FinishedGameStatisticsMapper.kt
package com.alextos.thousand.data.repository.mappers

import com.alextos.thousand.domain.models.Player
import com.alextos.thousand.domain.models.RemoteGame
import com.alextos.thousand.domain.models.RemoteUserInfo
import com.google.firebase.database.ServerValue

private const val USERS_NODE = "users"
private const val TOTAL_GAMES_COUNT_NODE = "gameCount"
private const val WIN_COUNT_NODE = "winCount"
private const val RATING_NODE = "rating"

internal fun RemoteGame.toFinishedGameStatisticsUpdates(
    userInfo: Map<String, RemoteUserInfo>
): Map<String, Any> {
    val updates = mutableMapOf<String, Any>()

    players.forEach { player ->
        val userId = player.user.id
        if (userId.isBlank()) {
            return@forEach
        }

        updates["$USERS_NODE/$userId/$TOTAL_GAMES_COUNT_NODE"] = ServerValue.increment(1)
        if (player.isWinner) {
            updates["$USERS_NODE/$userId/$WIN_COUNT_NODE"] = ServerValue.increment(1)
        }

        userInfo[player.user.id]?.let { remoteUserInfo ->
            updates["$USERS_NODE/$userId/$RATING_NODE"] = remoteUserInfo.rating
        }
    }

    return updates
}

File: iosApp/iosApp/Helpers/mappers/FinishedGameStatisticsMapper.swift
import ComposeApp
import FirebaseDatabase

extension IOSMultiplayerRepository {
    func finishedGameStatisticsUpdates(
        for game: RemoteGame,
        userInfo: [String: RemoteUserInfo]
    ) -> [String: Any] {
        var updates: [String: Any] = [:]

        game.players.forEach { player in
            let userId = player.user.id
            if userId.isEmpty {
                return
            }

            updates[FirebasePath.userField(userId: userId, field: FirebasePath.gameCount)] = ServerValue.increment(1)
            if player.isWinner {
                updates[FirebasePath.userField(userId: userId, field: FirebasePath.winCount)] = ServerValue.increment(1)
            }

            if let remoteUserInfo = userInfo[userId] {
                updates[FirebasePath.userField(userId: userId, field: FirebaseUserKey.rating)] = remoteUserInfo.rating
            }
        }

        return updates
    }
}
